File: java/sortcolors.java
public class Solution {
    public void sortColors(int[] A) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(A.length<=1) return;
        int a1,a2,a3;
        a1 = a2 = a3 = 0;
        a2 = a3 = A.length-1;
        
        while(a1<=a2){
            if(A[a1]==0){
                a1++;
                continue;
            }
            if(A[a1]==2){
                A[a1] = A[a3];
                A[a3] = 2;
                a3--;
                if(a2>a3){
                    a2 = a3;
                }
                continue;
            }
            if(A[a1]==1){
                A[a1] = A[a2];
                A[a2] = 1;
                a2--;
                continue;
            }
        }
    }
}

File: cpp/SortColors.cpp
class Solution {
public:
    void sortColors(int A[], int n) {
        int zeroBoundary = 0; // < zeroBoundary is 0
        int twoBoundary = n-1; // > twoBOundary is 2
        int start = 0;
        while(start <= twoBoundary) {
            if(A[start] == 0) {
                A[start] = A[zeroBoundary];
                A[zeroBoundary] = 0;
                zeroBoundary++;
                if(zeroBoundary>start) {
                    start = zeroBoundary;
                }
            } else if(A[start] == 2) {
                A[start] = A[twoBoundary];
                A[twoBoundary] = 2;
                twoBoundary--;
            } else {
                start++;
            }
        }
    }
};
